File: action_ws/src/balance_control/scripts/robot_command_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
========================================
Balance Control Command Client
========================================

功能说明:
  - 提供高层运动命令接口 (前进、后退、转向、跳跃等)
  - 将高层命令转换为底层平衡控制目标
  - 支持持续发送和一次性命令

命令类型:
  - STOP (0): 停止移动
  - FORWARD (1): 前进
  - BACKWARD (2): 后退
  - TURN_LEFT (3): 左转
  - TURN_RIGHT (4): 右转
  - STRAFE_LEFT (5): 左侧向移动
  - STRAFE_RIGHT (6): 右侧向移动
  - JUMP (7): 跳跃
  - BALANCE (8): 进入平衡模式

使用示例:
  rosrun balance_control robot_command_client.py _command:=1 _speed:=0.5
  rosrun balance_control robot_command_client.py _command:=3 _yaw_rate:=1.0

Author: Qi Xiao
Email: 2408128687@qq.com
日期: 2026-05-06
"""

import rospy
import actionlib
from balance_control.msg import RobotCommand
from balance_control.msg import BalanceControlAction, BalanceControlGoal
from geometry_msgs.msg import Twist

class RobotCommandClient:
    """
    机器人命令客户端类
    
    功能:
      - 订阅 /cmd_vel 话题获取运动命令
      - 将运动命令转换为平衡控制目标
      - 通过 Action 接口发送控制目标
    
    坐标系 (B frame):
      - X轴: 指向机器人右侧
      - Y轴: 指向机器人后方 (前进方向为-Y)
      - Z轴: 指向上方
    """

    def __init__(self):
        """初始化ROS节点和Action Client"""
        
        # ========== ROS初始化 ==========
        rospy.init_node('robot_command_client', anonymous=True)
        rospy.loginfo("==========================================")
        rospy.loginfo("Robot Command Client 已启动")
        rospy.loginfo("==========================================")
        
        # ========== 参数获取 ==========
        self.command = rospy.get_param('~command', 0)      # 默认停止
        self.speed = rospy.get_param('~speed', 0.0)        # 默认速度 0
        self.yaw_rate = rospy.get_param('~yaw_rate', 0.0)  # 默认转向 0
        self.jump_height = rospy.get_param('~jump_height', 0.0)
        self.duration = rospy.get_param('~duration', 0.0)
        
        # 持续发送模式 (用于cmd_vel)
        self.continuous_mode = rospy.get_param('~continuous', False)
        
        # ========== 关节映射 ==========
        # 左腿关节顺序: HIP_ROLL, HIP_PITCH, KNEE_PITCH, WHEEL
        # 右腿关节顺序: HIP_ROLL, HIP_PITCH, KNEE_PITCH, WHEEL
        self.joint_names = [
            'joint_left_leg_1', 'joint_left_leg_2', 'joint_left_leg_3', 'joint_left_leg_4',
            'joint_right_leg_1', 'joint_right_leg_2', 'joint_right_leg_3', 'joint_right_leg_4'
        ]
        
        # ========== 创建Action Client ==========
        self.action_client = actionlib.SimpleActionClient('balance_control', BalanceControlAction)
        rospy.loginfo("等待 Balance Control Action Server...")
        
        if not self.action_client.wait_for_server(rospy.Duration(10.0)):
            rospy.logerr("无法连接到 Action Server!")
            return
        
        rospy.loginfo("已连接到 Action Server")
        
        # ========== 订阅cmd_vel话题 ==========
        if self.continuous_mode:
            self.cmd_vel_sub = rospy.Subscriber('/cmd_vel', Twist, self.cmdVelCallback)
            rospy.loginfo("订阅 /cmd_vel 话题 (持续模式)")
        
        # ========== 发送初始命令 ==========
        self.sendCommand(self.command, self.speed, self.yaw_rate, self.jump_height, self.duration)
        
        # ========== 如果不是持续模式，则完成后退出 ==========
        if not self.continuous_mode:
            rospy.loginfo("命令已发送，客户端将退出")
            return
        
        # ========== 持续模式主循环 ==========
        rospy.loginfo("进入持续控制模式...")
        rate = rospy.Rate(10)  # 10Hz
        
        while not rospy.is_shutdown():
            rate.sleep()
            
            if rospy.is_shutdown():
                break
        
        # 退出前发送停止命令
        self.sendStop()
        rospy.loginfo("客户端退出")

    def cmdVelCallback(self, msg):
        """
        cmd_vel 回调函数
        
        将 Twist 消息转换为平衡控制命令:
          - linear.x -> 前进/后退 (Y方向)
          - linear.y -> 侧向移动 (X方向)
          - angular.z -> 转向 (Yaw)
        """
        # 从 cmd_vel 提取速度
        forward_speed = -msg.linear.x  # 转换为机器人坐标系
        lateral_speed = msg.linear.y
        yaw_rate = msg.angular.z
        
        # 判断命令类型
        command = RobotCommand.CMD_FORWARD  # 默认前进
        
        # 根据速度判断命令
        if abs(forward_speed) > 0.01:
            if forward_speed > 0:
                command = RobotCommand.CMD_FORWARD
            else:
                command = RobotCommand.CMD_BACKWARD
        elif abs(lateral_speed) > 0.01:
            if lateral_speed > 0:
                command = RobotCommand.CMD_STRAFE_RIGHT
            else:
                command = RobotCommand.CMD_STRAFE_LEFT
        elif abs(yaw_rate) > 0.01:
            if yaw_rate > 0:
                command = RobotCommand.CMD_TURN_LEFT
            else:
                command = RobotCommand.CMD_TURN_RIGHT
        else:
            command = RobotCommand.CMD_STOP
        
        speed = max(abs(forward_speed), abs(lateral_speed))
        self.sendCommand(command, speed, yaw_rate, 0.0, 0.0)

    def computeTargetOrientation(self, command, speed, yaw_rate):
        """
        根据命令计算目标姿态
        
        说明:
          - 前进/后退时，身体微微前倾/后仰
          - 转向时，身体微微侧倾
          - 保持平衡
        
        参数:
          command: 命令类型
          speed: 速度 m/s
          yaw_rate: 转向角速度 rad/s
        
        返回:
          (target_roll, target_pitch, target_yaw)
        """
        # 默认目标 (保持平衡)
        target_roll = 0.0
        target_pitch = 0.0
        target_yaw = 0.0
        
        if command == RobotCommand.CMD_FORWARD:
            # 前进: 身体微微前倾，角度与速度成正比
            target_pitch = -speed * 0.15  # 最大约 -0.3 rad (约17度)
        
        elif command == RobotCommand.CMD_BACKWARD:
            # 后退: 身体微微后仰
            target_pitch = speed * 0.15
        
        elif command == RobotCommand.CMD_TURN_LEFT:
            # 左转: 身体微微右倾，帮助转向
            target_roll = yaw_rate * 0.1
            target_yaw = yaw_rate * 0.5  # 直接控制偏航
        
        elif command == RobotCommand.CMD_TURN_RIGHT:
            # 右转: 身体微微左倾
            target_roll = -yaw_rate * 0.1
            target_yaw = -yaw_rate * 0.5
        
        elif command == RobotCommand.CMD_STRAFE_LEFT:
            # 左侧向: 身体微微右倾
            target_roll = 0.1
        
        elif command == RobotCommand.CMD_STRAFE_RIGHT:
            # 右侧向: 身体微微左倾
            target_roll = -0.1
        
        elif command == RobotCommand.CMD_STOP:
            # 停止: 回到中立姿态
            target_roll = 0.0
            target_pitch = 0.0
            target_yaw = 0.0
        
        elif command == RobotCommand.CMD_JUMP:
            # 跳跃: 身体微微下蹲准备
            target_pitch = -0.2
        
        elif command == RobotCommand.CMD_BALANCE:
            # 平衡模式: 保持当前姿态
            pass
        
        return target_roll, target_pitch, target_yaw

    def sendCommand(self, command, speed, yaw_rate, jump_height, duration):
        """
        发送命令到平衡控制器
        
        参数:
          command: 命令类型 (0-8)
          speed: 速度 m/s
          yaw_rate: 转向角速度 rad/s
          jump_height: 跳跃高度 m
          duration: 持续时间 s
        """
        # 计算目标姿态
        target_roll, target_pitch, target_yaw = self.computeTargetOrientation(command, speed, yaw_rate)
        
        # 创建 Action Goal
        goal = BalanceControlGoal()
        goal.command = 4  # CMD_SET_TARGET
        goal.algorithm = 0  # 0=LQR, 1=VMC, 2=MPC
        goal.target_roll = target_roll
        goal.target_pitch = target_pitch
        goal.target_yaw = target_yaw
        
        # 发送目标
        self.action_client.send_goal(goal)
        
        # 打印命令信息
        cmd_name = self.getCommandName(command)
        rospy.loginfo("==========================================")
        rospy.loginfo("发送命令: %s", cmd_name)
        rospy.loginfo("  速度: %.2f m/s", speed)
        rospy.loginfo("  转向: %.2f rad/s", yaw_rate)
        rospy.loginfo("  目标姿态: roll=%.3f, pitch=%.3f, yaw=%.3f", 
                      target_roll, target_pitch, target_yaw)
        rospy.loginfo("==========================================")
        
        # 如果不是持续命令，等待结果
        if duration > 0:
            rospy.sleep(duration)
            self.sendStop()

    def sendStop(self):
        """发送停止命令"""
        goal = BalanceControlGoal()
        goal.command = 2  # CMD_DISABLE
        self.action_client.send_goal(goal)
        rospy.loginfo("发送停止命令")

    def getCommandName(self, command):
        """获取命令名称"""
        names = {
            0: "STOP",
            1: "FORWARD",
            2: "BACKWARD",
            3: "TURN_LEFT",
            4: "TURN_RIGHT",
            5: "STRAFE_LEFT",
            6: "STRAFE_RIGHT",
            7: "JUMP",
            8: "BALANCE"
        }
        return names.get(command, "UNKNOWN")


if __name__ == '__main__':
    try:
        RobotCommandClient()
    except rospy.ROSInterruptException:
        pass

File: action_ws/src/balance_control/src/robot_command_client.cpp
/*
 * Robot Command Client Implementation
 * ===================================
 * 
 * Author: Qi Xiao
Email: 2408128687@qq.com
 * 日期: 2026-05-06
 */

#include "robot_command_client.h"
#include <ros/ros.h>

namespace balance_control {

RobotCommandClient::RobotCommandClient(ros::NodeHandle& nh)
    : nh_(nh)
    , action_client_("balance_control", true)
    , command_(RobotCommandType::STOP)
    , speed_(0.0)
    , yaw_rate_(0.0)
    , jump_height_(0.0)
    , duration_(0.0)
    , continuous_mode_(false)
    , cmd_vel_enabled_(false)
{
    // ========== 获取参数 ==========
    int command_int = 0;
    nh.param("command", command_int, 0);
    command_ = static_cast<RobotCommandType>(command_int);
    nh.param("speed", speed_, 0.0);
    nh.param("yaw_rate", yaw_rate_, 0.0);
    nh.param("jump_height", jump_height_, 0.0);
    nh.param("duration", duration_, 0.0);
    nh.param("continuous", continuous_mode_, false);

    // ========== 初始化关节配置 ==========
    joint_configs_.push_back({"joint_left_leg_1", 0, 0});
    joint_configs_.push_back({"joint_left_leg_2", 0, 1});
    joint_configs_.push_back({"joint_left_leg_3", 0, 2});
    joint_configs_.push_back({"joint_left_leg_4", 0, 3});
    joint_configs_.push_back({"joint_right_leg_1", 1, 0});
    joint_configs_.push_back({"joint_right_leg_2", 1, 1});
    joint_configs_.push_back({"joint_right_leg_3", 1, 2});
    joint_configs_.push_back({"joint_right_leg_4", 1, 3});

    // ========== 等待Action Server ==========
    ROS_INFO("等待 Balance Control Action Server...");

    if (!action_client_.waitForServer(ros::Duration(10.0))) {
        ROS_ERROR("无法连接到 Action Server!");
        return;
    }

    ROS_INFO("已连接到 Action Server");
    ROS_INFO("==========================================");

    // ========== 如果是持续模式，订阅cmd_vel ==========
    if (continuous_mode_) {
        cmd_vel_sub_ = nh.subscribe("/cmd_vel", 10,
                                    &RobotCommandClient::cmdVelCallback, this);
        cmd_vel_enabled_ = true;
        ROS_INFO("订阅 /cmd_vel 话题 (持续模式)");
    }

    // ========== 发送初始命令 ==========
    sendCommand(command_, speed_, yaw_rate_, jump_height_, duration_);

    // ========== 如果不是持续模式，则完成后退出 ==========
    if (!continuous_mode_) {
        ROS_INFO("命令已发送，客户端将退出");
        return;
    }

    // ========== 持续模式主循环 ==========
    ROS_INFO("进入持续控制模式...");
    ros::Rate rate(10);  // 10Hz

    while (ros::ok()) {
        ros::spinOnce();
        rate.sleep();
    }

    // 退出前发送停止命令
    sendStop();
    ROS_INFO("客户端退出");
}

RobotCommandClient::~RobotCommandClient() {
    if (cmd_vel_enabled_) {
        cmd_vel_sub_.shutdown();
    }
}

void RobotCommandClient::cmdVelCallback(const geometry_msgs::Twist::ConstPtr& msg) {
    // 从 cmd_vel 提取速度
    // 注意: 机器人坐标系中前进方向是-Y
    double forward_speed = -msg->linear.x;
    double lateral_speed = msg->linear.y;
    double yaw_rate = msg->angular.z;

    // 判断命令类型
    RobotCommandType command = RobotCommandType::FORWARD;
    double speed = 0.0;

    if (std::abs(forward_speed) > 0.01) {
        command = (forward_speed > 0) ? RobotCommandType::FORWARD : RobotCommandType::BACKWARD;
        speed = std::abs(forward_speed);
    } else if (std::abs(lateral_speed) > 0.01) {
        command = (lateral_speed > 0) ? RobotCommandType::STRAFE_RIGHT : RobotCommandType::STRAFE_LEFT;
        speed = std::abs(lateral_speed);
    } else if (std::abs(yaw_rate) > 0.01) {
        command = (yaw_rate > 0) ? RobotCommandType::TURN_LEFT : RobotCommandType::TURN_RIGHT;
        speed = std::abs(yaw_rate);
    } else {
        command = RobotCommandType::STOP;
        speed = 0.0;
    }

    sendCommand(command, speed, yaw_rate, 0.0, 0.0);
}

TargetOrientation RobotCommandClient::computeTargetOrientation(
    RobotCommandType command, double speed, double yaw_rate) {

    TargetOrientation target = {0.0, 0.0, 0.0};

    switch (command) {
        case RobotCommandType::FORWARD:
            // 前进: 身体微微前倾，角度与速度成正比
            target.pitch = -speed * 0.15;  // 最大约 -0.3 rad (约17度)
            break;

        case RobotCommandType::BACKWARD:
            // 后退: 身体微微后仰
            target.pitch = speed * 0.15;
            break;

        case RobotCommandType::TURN_LEFT:
            // 左转: 身体微微右倾，帮助转向
            target.roll = yaw_rate * 0.1;
            target.yaw = yaw_rate * 0.5;  // 直接控制偏航
            break;

        case RobotCommandType::TURN_RIGHT:
            // 右转: 身体微微左倾
            target.roll = -yaw_rate * 0.1;
            target.yaw = -yaw_rate * 0.5;
            break;

        case RobotCommandType::STRAFE_LEFT:
            // 左侧向: 身体微微右倾
            target.roll = 0.1;
            break;

        case RobotCommandType::STRAFE_RIGHT:
            // 右侧向: 身体微微左倾
            target.roll = -0.1;
            break;

        case RobotCommandType::STOP:
            // 停止: 回到中立姿态
            target.roll = 0.0;
            target.pitch = 0.0;
            target.yaw = 0.0;
            break;

        case RobotCommandType::JUMP:
            // 跳跃: 身体微微下蹲准备
            target.pitch = -0.2;
            break;

        case RobotCommandType::BALANCE:
            // 平衡模式: 保持当前姿态
            break;
    }

    return target;
}

void RobotCommandClient::sendCommand(RobotCommandType command, double speed,
                                     double yaw_rate, double jump_height, double duration) {
    // 计算目标姿态
    TargetOrientation target = computeTargetOrientation(command, speed, yaw_rate);

    // 创建 Action Goal
    balance_control::BalanceControlGoal goal;
    goal.enable_control = true;
    goal.algorithm_id = 0;  // 0=LQR, 1=VMC, 2=MPC
    goal.target_roll = target.roll;
    goal.target_pitch = target.pitch;
    goal.target_yaw = target.yaw;

    // 发送目标
    action_client_.sendGoal(goal);

    // 打印命令信息
    ROS_INFO("==========================================");
    ROS_INFO("发送命令: %s", commandToString(command).c_str());
    ROS_INFO("  速度: %.2f m/s", speed);
    ROS_INFO("  转向: %.2f rad/s", yaw_rate);
    ROS_INFO("  目标姿态: roll=%.3f, pitch=%.3f, yaw=%.3f",
             target.roll, target.pitch, target.yaw);
    ROS_INFO("==========================================");

    // 如果不是持续命令，等待结果
    if (duration > 0) {
        ros::Duration(duration).sleep();
        sendStop();
    }
}

void RobotCommandClient::sendStop() {
    balance_control::BalanceControlGoal goal;
    goal.enable_control = false;
    action_client_.sendGoal(goal);
    ROS_INFO("发送停止命令");
}

void RobotCommandClient::sendEnable() {
    balance_control::BalanceControlGoal goal;
    goal.enable_control = true;
    action_client_.sendGoal(goal);
    ROS_INFO("发送启用命令");
}

void RobotCommandClient::sendTarget(double roll, double pitch, double yaw) {
    balance_control::BalanceControlGoal goal;
    goal.enable_control = true;
    goal.target_roll = roll;
    goal.target_pitch = pitch;
    goal.target_yaw = yaw;
    action_client_.sendGoal(goal);
    ROS_INFO("发送目标姿态: roll=%.3f, pitch=%.3f, yaw=%.3f", roll, pitch, yaw);
}

void RobotCommandClient::setCmdVelEnabled(bool enabled) {
    cmd_vel_enabled_ = enabled;
    if (enabled && !cmd_vel_sub_) {
        cmd_vel_sub_ = nh_.subscribe("/cmd_vel", 10,
                                     &RobotCommandClient::cmdVelCallback, this);
    }
}

std::string RobotCommandClient::commandToString(RobotCommandType command) {
    switch (command) {
        case RobotCommandType::STOP: return "STOP";
        case RobotCommandType::FORWARD: return "FORWARD";
        case RobotCommandType::BACKWARD: return "BACKWARD";
        case RobotCommandType::TURN_LEFT: return "TURN_LEFT";
        case RobotCommandType::TURN_RIGHT: return "TURN_RIGHT";
        case RobotCommandType::STRAFE_LEFT: return "STRAFE_LEFT";
        case RobotCommandType::STRAFE_RIGHT: return "STRAFE_RIGHT";
        case RobotCommandType::JUMP: return "JUMP";
        case RobotCommandType::BALANCE: return "BALANCE";
        default: return "UNKNOWN";
    }
}

} // namespace balance_control

/*
 * Main Entry Point
 * ================
 */
int main(int argc, char** argv) {
    ros::init(argc, argv, "robot_command_client");

    try {
        ros::NodeHandle nh;
        balance_control::RobotCommandClient client(nh);
    } catch (const std::exception& e) {
        ROS_ERROR("异常: %s", e.what());
        return 1;
    }

    return 0;
}